### PR TITLE
SDK-921. Color labels and favorite files / folders

### PR DESCRIPTION
**Pull request by jgandres**

### CODE REVIEW

File: include/megaapi.h [hunk 1]
```diff
@@ -10104,6 +10127,55 @@ class MegaApi
          */
         void setNodeDuration(MegaNode *node, int duration, MegaRequestListener *listener = NULL);
 
+        /**
```
[the remark this answers is not in the document]
Comment (jgandres) [reply include/megaapi.h:10141]: Fixed in 1fea92b35488a65eddda6dddd2af64f106cbf53b

File: src/megaapi_impl.cpp [hunk 2]
```diff
@@ -170,6 +174,8 @@ MegaNodePrivate::MegaNodePrivate(MegaNode *node)
         this->height = node->getHeight();
         this->shortformat = node->getShortformat();
         this->videocodecid = node->getVideocodecid();
+        this->favourite = node->isFavourite();
+        this->label = static_cast<nodelabel_t>(node->getLabel());
```
Comment (sergiohs84) [inline src/megaapi_impl.cpp:178]: IIRC, in the ctor you can access to the members directly (avoiding the casting, in this case)
```suggestion
        this->label = node->label;
```

File: src/megaapi_impl.cpp [hunk 3]
```diff
@@ -390,6 +398,30 @@ MegaNodePrivate::MegaNodePrivate(Node *node)
...
+            else if (it->first == AttrMap::string2nameid("fav"))
+            {
+                int fav = std::atoi(it->second.c_str());
+                if (fav != 1)
+                {
+                    LOG_err << "Invalid value for node attr fav: " << fav;
+                }
+                else
+                {
+                    favourite = std::atoi(it->second.c_str());
```
Comment (sergiohs84) [inline src/megaapi_impl.cpp:410]:
```suggestion
                    favourite = fav;
```

File: src/megaapi_impl.cpp [hunk 4]
```diff
@@ -21936,6 +22058,25 @@ void MegaApiImpl::sendPendingRequests()
...
+                   attrs.map[favnid] = it->second;
+               }
+
+               nameid lblnid = AttrMap::string2nameid("lbl");
+               it = previousNode->attrs.map.find(lblnid);
+               if (it != previousNode->attrs.map.end())
+               {
+                   attrs.map[lblnid] = it->second;
+               }
+            }
```
Comment (sergiohs84) [inline src/megaapi_impl.cpp:22078]: Isn't this block duplicated? If so, perhaps it's better to encapsulate the code in a method, and use it wherever it's needed.

File: src/megaapi_impl.cpp (hunk 2 above)
Comment (jgandres) [reply src/megaapi_impl.cpp]: MegaNode::getLabel returns int to ensure it's compatible with bindings, and label type is nodelabel_t, so you need a casting to make the assignation

File: src/megaapi_impl.cpp (hunk 4 above)
Comment (jgandres) [reply src/megaapi_impl.cpp]: Done in commit: ec73f8b66879e3437fecdf3942815196b147d013

File: include/megaapi.h [hunk 5]
```diff
@@ -10104,6 +10138,55 @@ class MegaApi
...
+         * current value of the attribute for the node.
+         *
+         * The associated request type with this request is MegaRequest::TYPE_SET_ATTR_NODE
+         * Valid data in the MegaRequest object received on callbacks:
+         * - MegaRequest::getNodeHandle - Returns the handle of the node that receive the attribute
+         * - MegaRequest::getNumDetails - Returns the label for the node
+         * - MegaRequest::getFlag - Returns true (official attribute)
+         * - MegaRequest::getParamType - Returns MegaApi::NODE_ATTR_LABEL
+         *
+         * If the MEGA account is a business account and it's status is expired, onRequestFinish will
```
[the remark this answers is not in the document]
Comment (jgandres) [reply include/megaapi.h]: I thought that business errors should be documented in every method, to make easier developers identify which actions are affected by API_EBUSINESSPASTDUE. Anyway I've removed it: 4c935fe3ebc26d3e2ba9efeb82782934ac8e0452

**Comment by jgandres** [comment]
trigger compilation

**Comment by jgandres** [comment]
trigger compilation

**Comment by sergiohs84** [comment]
trigger compilation android

**Review by sergiohs84 — APPROVED** [review]

**Review by sergiohs84 — DISMISSED** [review]

**Review by alber2510 — DISMISSED** [review]

**Comment by jgandres** [comment]
Changes related to order by label/favourite in PR: https://github.com/meganz/sdk/pull/2242

**Review by sergiohs84 — APPROVED** [review]

**Pull request merged**